src/java/org/apache/cassandra/index/sai/StorageAttachedIndexGroup.java

```diff
@@ -250,31 +261,66 @@ public boolean handles(IndexTransaction.Type type)
     }
 
     @Override
-    public Set<Component> getComponents()
-    {
-        return getComponents(indices);
-    }
+    public Set<Component> componentsForNewBuid(Descriptor descriptor, TableMetadata metadata)
+    {
+        return componentsForNewBuid(descriptor, metadata, indices);
+    }
+
+    static Set<Component> componentsForNewBuid(Descriptor descriptor, TableMetadata metadata, Collection<StorageAttachedIndex> indices)
+    {
+        // The components created depends, amongst other things, on the generation that needs to be used for the new
+        // build (which depends on the version, but that's always the latest for new builds). That generation is always
+        // 0 for new sstables, but may not be for rebuilds. To handle that, we rely on the same logic used when we do
+        // write the new components (`IndexDescriptor.newPerSSTableGroupWriter`/`IndexDescriptor.newPerIndexGroupWriter`),
+        // even if we don't write anything at that point.
+        // Do note that we create an `IndexDescriptor` from scratch, rather than getting the one from the context
+        // manager, because in some cases, the underlying sstable will simply not exist yet (and when that's the case,
+        // we don't want to populate the context just yet). It will exist if this is called as part of a rebuild or is
+        // the build of a new index on an existing sstable, but this is also called for sstables that we're about to
+        // flush or about to create as result of compaction. Of course, we could it from the context in the case of
+        // existing stables, and create it otherwise, but it doesn't feel worth bothering here.
+        IndexDescriptor indexDescriptor = IndexDescriptor.create(descriptor, metadata);
+        Set<Component> components = indexDescriptor
+                                    .newPerSSTableComponentsForWrite()
```

Comment:
Would be nice to have a test in CNDB to verify index build after index drop: 
- When an index is dropped on CNDB, writer deletes index files locally without updating archive file, so the zipped TOC remains the same. If the user creates an index again, compactor should still see the old TOC with old index component generation and should use next generation.

Reply:
Yeah. There is test for drop+recreate for the NSS case, so that will be run once we merge this to cndb, but I don't know about the non-NSS case. I'll check if there is one once I merge this and propagate the change, and I'll add a test if that's not the case to make sure it's not broken.